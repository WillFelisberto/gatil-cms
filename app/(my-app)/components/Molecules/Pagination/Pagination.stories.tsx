import type { Meta, StoryObj } from '@storybook/react';

import { Pagination } from './Pagination';

const meta: Meta<typeof Pagination> = {
  title: 'Molecules/Pagination',
  component: Pagination,
  tags: ['autodocs'],
  argTypes: {
    currentPage: {
      control: { type: 'number', min: 1 },
      defaultValue: 1
    },
    totalPages: {
      control: { type: 'number', min: 1 },
      defaultValue: 5
    }
  }
};

export default meta;
type Story = StoryObj<typeof Pagination>;

export const Default: Story = {
  args: {
    currentPage: 1,
    totalPages: 5
  }
};
