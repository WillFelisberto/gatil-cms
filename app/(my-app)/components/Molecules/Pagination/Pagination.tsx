'use client';

import { ArrowLeft, ArrowRight } from 'lucide-react';
import { useRouter, useSearchParams } from 'next/navigation';

type PaginationProps = {
  currentPage: number;
  totalPages: number;
};

export const Pagination = ({ currentPage, totalPages }: PaginationProps) => {
  const router = useRouter();
  const searchParams = useSearchParams();

  const goToPage = (page: number) => {
    const params = new URLSearchParams(searchParams);
    if (page === 1) {
      params.delete('page');
    } else {
      params.set('page', page.toString());
    }
    router.push(`?${params.toString()}`);
  };

  return (
    <nav
      className="flex gap-2 items-center flex-wrap justify-center mt-8"
      role="navigation"
      aria-label="Paginação de gatinhos"
      data-testid="pagination"
    >
      <button
        disabled={currentPage === 1}
        onClick={() => goToPage(currentPage - 1)}
        className="p-2 rounded-full border border-[#013274] text-[#013274] hover:bg-[#013274]/10 disabled:opacity-30 transition"
        aria-label="Página anterior"
        data-testid="prev-button"
      >
        <ArrowLeft className="w-4 h-4" />
      </button>

      {Array.from({ length: totalPages }, (_, i) => i + 1).map((page) => (
        <button
          key={page}
          onClick={() => goToPage(page)}
          className={`w-9 h-9 rounded-full border flex items-center justify-center transition font-medium ${
            page === currentPage
              ? 'bg-[#013274] text-white border-[#013274] shadow-md'
              : 'border-[#013274] text-[#013274] hover:bg-[#013274]/10'
          }`}
          aria-current={page === currentPage ? 'page' : undefined}
          aria-label={`Página ${page}`}
          data-testid={`page-button-${page}`}
        >
          {page}
        </button>
      ))}

      <button
        disabled={currentPage === totalPages}
        onClick={() => goToPage(currentPage + 1)}
        className="p-2 rounded-full border border-[#013274] text-[#013274] hover:bg-[#013274]/10 disabled:opacity-30 transition"
        aria-label="Próxima página"
        data-testid="next-button"
      >
        <ArrowRight className="w-4 h-4" />
      </button>
    </nav>
  );
};
