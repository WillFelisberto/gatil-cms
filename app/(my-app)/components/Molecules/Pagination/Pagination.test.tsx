import * as nextNavigation from 'next/navigation';

import { fireEvent, render, screen } from '@/tests/test-utils';

import { Pagination } from './Pagination';

// mock de router.push
jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
  useSearchParams: jest.fn()
}));

describe('<Pagination />', () => {
  const pushMock = jest.fn();
  beforeEach(() => {
    jest.clearAllMocks();

    (nextNavigation.useRouter as jest.Mock).mockReturnValue({
      push: pushMock
    });

    (nextNavigation.useSearchParams as jest.Mock).mockReturnValue(new URLSearchParams('page=2'));
  });

  it('should render pagination with correct buttons', () => {
    const { container } = render(<Pagination currentPage={2} totalPages={4} />);

    expect(screen.getByTestId('pagination')).toBeInTheDocument();
    expect(screen.getByTestId('prev-button')).toBeInTheDocument();
    expect(screen.getByTestId('next-button')).toBeInTheDocument();

    for (let i = 1; i <= 4; i++) {
      expect(screen.getByTestId(`page-button-${i}`)).toBeInTheDocument();
    }

    expect(container).toMatchSnapshot();
  });

  it('should disable previous button when on first page', () => {
    render(<Pagination currentPage={1} totalPages={3} />);
    expect(screen.getByTestId('prev-button')).toBeDisabled();
    expect(screen.getByTestId('next-button')).not.toBeDisabled();
  });

  it('should disable next button when on last page', () => {
    render(<Pagination currentPage={3} totalPages={3} />);
    expect(screen.getByTestId('next-button')).toBeDisabled();
    expect(screen.getByTestId('prev-button')).not.toBeDisabled();
  });

  it('should highlight current page', () => {
    render(<Pagination currentPage={2} totalPages={5} />);
    expect(screen.getByTestId('page-button-2')).toHaveAttribute('aria-current', 'page');
  });

  it('should navigate to next page on next button click', () => {
    render(<Pagination currentPage={2} totalPages={5} />);
    fireEvent.click(screen.getByTestId('next-button'));
    expect(pushMock).toHaveBeenCalledWith('?page=3');
  });

  it('should navigate to previous page on previous button click', () => {
    render(<Pagination currentPage={2} totalPages={5} />);
    fireEvent.click(screen.getByTestId('prev-button'));
    expect(pushMock).toHaveBeenCalledWith('?');
  });

  it('should navigate to specific page on page button click', () => {
    render(<Pagination currentPage={1} totalPages={5} />);
    fireEvent.click(screen.getByTestId('page-button-4'));
    expect(pushMock).toHaveBeenCalledWith('?page=4');
  });
});
