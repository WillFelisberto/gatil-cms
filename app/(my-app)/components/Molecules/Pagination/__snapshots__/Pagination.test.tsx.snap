// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> should render pagination with correct buttons 1`] = `
<div>
  <nav
    aria-label="Paginação de gatinhos"
    class="flex gap-2 items-center flex-wrap justify-center mt-8"
    data-testid="pagination"
    role="navigation"
  >
    <button
      aria-label="Página anterior"
      class="p-2 rounded-full border border-[#013274] text-[#013274] transition disabled:opacity-30 hover:bg-[#013274]/10 cursor-pointer"
      data-testid="prev-button"
    >
      <svg
        class="lucide lucide-arrow-left w-4 h-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m12 19-7-7 7-7"
        />
        <path
          d="M19 12H5"
        />
      </svg>
    </button>
    <button
      aria-label="Página 1"
      class="w-9 h-9 rounded-full border flex items-center justify-center transition font-medium border-[#013274] text-[#013274] hover:bg-[#013274]/10 cursor-pointer"
      data-testid="page-button-1"
    >
      1
    </button>
    <button
      aria-current="page"
      aria-label="Página 2"
      class="w-9 h-9 rounded-full border flex items-center justify-center transition font-medium bg-[#013274] text-white border-[#013274] shadow-md"
      data-testid="page-button-2"
    >
      2
    </button>
    <button
      aria-label="Página 3"
      class="w-9 h-9 rounded-full border flex items-center justify-center transition font-medium border-[#013274] text-[#013274] hover:bg-[#013274]/10 cursor-pointer"
      data-testid="page-button-3"
    >
      3
    </button>
    <button
      aria-label="Página 4"
      class="w-9 h-9 rounded-full border flex items-center justify-center transition font-medium border-[#013274] text-[#013274] hover:bg-[#013274]/10 cursor-pointer"
      data-testid="page-button-4"
    >
      4
    </button>
    <button
      aria-label="Próxima página"
      class="p-2 rounded-full border border-[#013274] text-[#013274] transition disabled:opacity-30 hover:bg-[#013274]/10 cursor-pointer"
      data-testid="next-button"
    >
      <svg
        class="lucide lucide-arrow-right w-4 h-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 12h14"
        />
        <path
          d="m12 5 7 7-7 7"
        />
      </svg>
    </button>
  </nav>
</div>
`;
